### DISCUSSION TITLE
Android support with use of WebUSB API

### DISCUSSION THREAD

**Discussion by kaklik** [Open-Gamma-Project/Gamma-MCA#20] [Show and tell]
Originally the Gamma-MCA does not support Android phones, because it uses the WebUSBSerial API and Android does not contain the necessary USB drivers.  (The same situation I expect in the case of Apple, but I have not tested it)

Therefore we made [a fork](https://github.com/UniversalScientificTechnologies/Gamma-MCA) where we added support for WebUSB. Demonstrated on [LABDOS semiconductor-based ionizing radiation spectrometer](https://github.com/UniversalScientificTechnologies/LABDOS01). That seems to actually allow to use of the Gamma-MCA on Android. 

![image](https://user-images.githubusercontent.com/1194177/216840080-1c52aa78-e126-4b18-b490-3c36a1de544e.png)

Is it that feature desirable in the Gamma-MCA?

**Comment by NuclearPhoenixx** [comment]
Very interesting! I wasn't aware of the fact that Web USB is supported on mobile. Sounds even more unnecessary that the Web Serial API is not then, but ok. It's all "experimental" APIs anyways and I think Safari and Firefox support neither. Web USB can be much more of a security vulnerability, though, so I get that.

I had a glance at the code and as far as I understand it, you are essentially adding support for FTDI(-like?) serial-to-USB adapters. You can create a PR if you want and I'll add my two cents ;)

**Reply by slimonslimon** [reply]
Im very nervous about security of this api, but the api is available, so we can use it.

One more ask. Should I cherrypick WebUSB support to extra commit before PR? Im asking, because we added special "serial input data type" for our device, and these modifications are mixed in commits. And this format of incoming data is only for testing now. We are planning (and our device i capable) send more robust format, which can mix both "histogram" and "event" data... Do you planning modification of your serial protocol?

**Reply by NuclearPhoenixx** [reply]
I don't plan on making modifications, but if you need to use your differential histogram, we can think about adding that. Would be much easier if you just switch the programming on your side, though. It would also add a lot of redundant code that is pretty similar to the already existing histogram mode. If you already have a differential histogram, using the standard histogram mode should not be too hard.

And yes, please try to make multiple PRs for each substantial change if you can. Would make reviewing the code way faster 👍🏻

**Reply by NuclearPhoenixx** [reply]
I forgot to mention that the best way to use the Web USB API is just to make it a fallback, only if Web Serial is not supported. This way we only expose the API if there is no way around it, i.e. on mobile.

**Reply by slimonslimon** [reply]
okey, it seem i didnt noticed other discussion threads.... I will prepare PR for WebUSB only.

**Comment by kaklik** [comment]
Ok, that seems to be solved by this pull request: https://github.com/OpenGammaProject/Gamma-MCA/pull/29

**Comment by NuclearPhoenixx** [comment]
Closing this discussion for now since this feature has been implemented now for quite some time using WebUSB and the common FTDI UART chips and it looks to work well.
